crates/walrus-service/src/node/pending_metadata_cache.rs

```diff
@@ -0,0 +1,145 @@
+// Copyright (c) Walrus Foundation
+// SPDX-License-Identifier: Apache-2.0
+
+//! In-memory cache for metadata awaiting blob registration.
+
+use std::{
+    collections::HashMap,
+    sync::Arc,
+    time::{Duration, Instant},
+};
+
+use tokio::sync::RwLock;
+use walrus_core::{BlobId, metadata::VerifiedBlobMetadataWithId};
+
+use super::metrics::NodeMetricSet;
+
+#[derive(Debug, Clone)]
+struct CachedMetadata {
+    inserted_at: Instant,
+    metadata: Arc<VerifiedBlobMetadataWithId>,
+}
+
+#[derive(Debug)]
+struct PendingMetadataCacheInner {
+    max_entries: usize,
+    ttl: Duration,
+    entries: HashMap<BlobId, CachedMetadata>,
+}
+
+impl PendingMetadataCacheInner {
+    fn new(max_entries: usize, ttl: Duration) -> Self {
+        Self {
+            max_entries,
+            ttl,
+            entries: HashMap::new(),
+        }
+    }
+
+    fn evict_expired(&mut self, now: Instant) {
+        if self.ttl.is_zero() {
+            return;
+        }
+        let ttl = self.ttl;
+        self.entries
+            .retain(|_, cached| now.saturating_duration_since(cached.inserted_at) < ttl);
+    }
+
+    fn get(&self, blob_id: &BlobId) -> Option<Arc<VerifiedBlobMetadataWithId>> {
+        self.entries
+            .get(blob_id)
+            .map(|cached| cached.metadata.clone())
+    }
+
+    /// Inserts metadata into the cache.
+    /// Returns `Err(())` when the cache is at capacity and the entry could not be stored after
+    /// evicting expired items.
+    fn insert(
+        &mut self,
+        blob_id: BlobId,
+        metadata: Arc<VerifiedBlobMetadataWithId>,
+    ) -> Result<(), ()> {
```

Comment:
Suggestion/nit: how about returning bool to indicating whether the metadata is added to the cache? It's a bit more explicit this way.

Reply:
done